sound/soc/sof/ipc4-pcm.c

```diff
@@ -21,12 +21,15 @@
  * @dai_copier: the dai copier of the pcm stream
  * @stream_start_offset: reported by fw in memory window
  * @llp_offset: llp offset in memory window
+ * @boundary: wrap boundary should be used for the LLP frame counter
```

Comment:
Commit message is really odd

"
To be able to calculate the current delay we need to use the DMA position
directly instead of the substream->runtime->status->hw_ptr, which is the
previous hw_pointer and it can be significantly different than where the
hardware currently is.
"

The way I read it is that with this PR we have an accurate delay calculation, but it's not related to the hw_ptr, so if an application does a calculation of DMA position + delay, then it won't be the sample currently playing...

I am struggling to understand the concept, the delay is a relative concept to the DMA position. Finding out the delay independently of the DMA position reported to userspace is a sure way to add uncertainty and measurement problems, no?

Reply:
@plbossart , OK, so the issue is two-fold, the use of `substream->runtime->status->hw_ptr,` was not correct as it is not the correct information needed for the delay reporting.

Then to the delay reporting:
when there is no delay (reported) it implies that the frame at the hw_ptr will be on the output at the next sample tick, right?
The delay tells the user space that the frame at the hw_ptr will _not_ be on the output until 'delay' number of frames worth of time passes, right?

It would be ideal if we could ask the firmware to tell us the delay, but it is not doing that, it only gives us two numbers: **stream_start_offset** and the frame counter on the dai side (**LLP**). This is not enough information, but how would the firmware - if it would - gave us a delay number?
When a stream is started it would calculate the **stream_start_offset** which is in essence the number of frame time it will take for the first sample copied to the DSP memory to be on the output.
The **LLP** is the number of frames left the DSP.
On stream start: we copied X frames to the DSP, it will take  Y frame time to reach the output:
when LLP=0 the delay is Y for the first frame, but the host needs to know the delay for the last frame that has been copied to the DSP (the hw_ptr is pointing to the next frame after this last one in the DSP after all), so for the delay that we report to user is: Y+X.
When LLP moves 1 frame, the delay is going to decrease: (Y-1)+X
When LLP moves 1 more frame: (Y-2)+X
...
When LLP moved 100 frames: (Y-100)+X
When LLP == X, that is the point when the first real frame is started to be played out, delay is now X
From this point the delay is: (LLP-Y)+X
and so on, the delay is the difference between the frames entered and left the DSP, adjusted with the stream_start_offset.

Since the firmware is not giving this information, we need to calculate it. We get the LLP and stream_start_offset (Y) from firmware but we need to get the X.
The X is _not_ the DMA pointer in the ALSA buffer, it has nothing to do with ALSA buffer. It is a counter between the host and DSP. Sure, the DMA is counting it but it is rather a counter than a pointer.
We just don't have other means of asking how much data entered or held within the DSP.
Other SoCs, codecs can tell you their FIFO occupation number (which they get by doing the same calculation of what entered and what left).

and why it is not appropriate to use the DMA position in the ALSA buffer?
Our topology tells alsa that the period_size_min is 192 (48 frames in stereo, 16bit - 1ms), the stream_start_offset is about 200 frames.
If application takes three 48 frame periods - which is allowed - then at start the delay is 200+96 frames (our DMA buffer is 2ms on the firmware side), that is hard to represent within the 144 frame long buffer...

The DMA pointer has nothing to do with the delay calculation.
The ALSA buffer has nothing to do with the delay calculation.

What matters:
The time that it takes for a sample to traverse within the DSP (stream_start_offset)
The number of frames left the DSP (LLP)
The number of frames entered the DSP (frame counter on host)

It is an ASoC thing that the pointer and delay calls got separated, but they are executed from the same ALSA pointer call, so yes, there can be some misalignment but that affects _all_ ASoC drivers.

Reply:
@plbossart, I think the main issue is that we don't really have a place for this to report, imho.

In essence we would need something somehow use  a counter on the BE and FE side and a start_offset, there is no API to get counter from neither side so we use the 'best' available place where we can do this calculation.

Reply:
I am not following your logic @ujfalusi 

My starting point is this: https://elixir.bootlin.com/linux/latest/source/sound/soc/soc-pcm.c#L1220

The soc_pcm_pointer() will get an offset and *IMMEDIATELY* after a delay.

I don't see the point of rechecking what the host-pointer is. 

We could get the delay at the same time as the pcm_ pointer, and store it in a cache variable and return it immediately when the snd_soc_pcm_component_delay is used.

the pointer() and delay() callbacks should not be seen as isolated functionality, ideally the pointer and delay should be computed at the same time.

Reply:
> > I am not following your logic @ujfalusi
> 
> @plbossart, OK, I got your point and understand it, but,
> 
> > My starting point is this: https://elixir.bootlin.com/linux/latest/source/sound/soc/soc-pcm.c#L1220
> 
> Yes, I know and for me that whole thing is a bit broken. it is taking the goes through the dais/components and returns the biggest delay reported by the dais/components checked. Imho it should be _adding_ the delays together. But that is another discussion...

The max() only applies to the case where there are multiple cpu or codec dais. It was selected because there's no good way of determining what the delay should be if multiple dais report different values.

> > I don't see the point of rechecking what the host-pointer is,
> 
> We don't check the host-pointer two times, that is the whole point. In `snd_soc_pcm_component_pointer()` we check the host-pointer via DPIB. The pointer is a DMA position within the ALSA buffer. In the delay callback we calculate the current delay the DSP causes, for which we use two counters and an offset value. DAI counter (LLP) and host counter (LDP) and stream_start_offset. LLP/LDP is counting the number of frames/bytes on the two side of the DSP.

Well you are using DPIB and then LDP. That's two ways of looking at the host pointer.
My question is whether we can use ONE register only.

> > and I don't understand what is calculated by the alsa core. In the ASoC case it looks rather simple: nothing is calculated by the alsa core....
> 
> Under ASoC we always have ALSA doing it's magic: https://elixir.bootlin.com/linux/latest/source/sound/core/pcm_lib.c#L283

Isn't this magic code applied AFTER doing the pointer/delay estimation in the platform driver?

> > We could get the delay at the same time as the pcm_ pointer, and store it in a cache variable and return it immediately when the snd_soc_pcm_component_delay is used.
> 
> In either way you cannot guaranty atomic operation. You still have two calls to get the pointer needed for the pointer and another to get the delay. The host DMA can move in between, it can move between taking the host counter and then the DAI counter or if you reverse them, the DAI counter has moved when you would take the host counter.
> 
> If we cache the delay value from taken at SOF pointer time and we have a codec reporting delay then again we have a DSP delay from the past with a current delay on the codec.
> 
> It is just that we cannot freeze the system to take the sample, it is moving.
> 
> Note: I did experiment similar to this, it was 'off' about the same frequency as the current implementation.
> 
> > the pointer() and delay() callbacks should not be seen as isolated functionality, ideally the pointer and delay should be computed at the same time.
> 
> For a platform driver, yes, likely it would be better but the DAI, codec, amp can also add delay and there is no way that you can stop everything and read them at the same time.

I was only trying to look at the component delay and avoid this DPIB/LDP dual use. Maybe it's a second order problem and maybe it doesn't bring any additional precision in practice, I don't know. It's just very odd to use two reference points instead of one.

Reply:
@plbossart, I don't want to use LDP for the pointer. We know that it works CAVS2.5+ but older arch?
The DPIB is convenient as it wraps in buffer.

But, in the core we only get the pointer which is wrapped within the buffer regardless if we derived it from DPIB or LDP. We need a counter.
Where we know the LDP we don't know the LLP (and stream_start_offset),

Either way we need to introduce the same new callbacks and just do the very same thing which we do in the delay callback in a new (we don't have ipc_ops for pointer as it was irrelevant what the DSP does - we cared about the host DMA in buffer). That might get us few usec closer to the pointer time readout of the counter but I don't see the benefit